First Pull request for review.

usman-altaf-co opened the pull request.

usman-altaf-co committed:
Add files via upload

usman-altaf-co committed:
Update weather.py

usman-altaf-co committed:
Update weather.py

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,220 @@
+
+months = ['Jan', 'Feb', 'Mar', 'Apr',
+          'May', 'Jun', 'Jul', 'Aug',
+          'Sep', 'Oct', 'Nov', 'Dec']
```
Can we search python utilities to manage months?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,220 @@
+          'May', 'Jun', 'Jul', 'Aug',
+          'Sep', 'Oct', 'Nov', 'Dec']
+
+labels = [
+    'PKT', 'Max TemperatureC', 'Mean TemperatureC',
+    'Min TemperatureC', 'Dew PointC', 'MeanDew PointC',
+    'Min DewpointC', 'Max Humidity', 'Mean Humidity',
+    'Min Humidity', 'Max Sea Level PressurehPa',
+    'Mean Sea Level PressurehPa', 'Min Sea Level PressurehPa',
+    'Max VisibilityKm', 'Mean VisibilityKm', 'Min VisibilitykM',
+    'Max Wind SpeedKm/h', 'Mean Wind SpeedKm/h', 'Max Gust SpeedKm/h',
+    'Precipitationmm', 'CloudCover', 'Events', 'WindDirDegrees'
+]
```
Do we need all these values for our work?

ahmed-buksh replied:
Comment still unaddressed!

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,220 @@
+    'Precipitationmm', 'CloudCover', 'Events', 'WindDirDegrees'
+]
+
+month_abbr = {
+    1: 'Jan', 2: 'Feb', 3: 'Mar',
+    4: 'Apr', 5: 'May', 6: 'Jun',
+    7: 'Jul', 8: 'Aug', 9: 'Sep',
+    10: 'Oct', 11: 'Nov', 12: 'Dec'
+}
```
Look into Python utilities for such basic work. No need of reinventing the wheel.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,220 @@
+
+                values = line.split(',')
+
+                # Dictionary Comprehension, Creating and Converting tuple into dictionary's key-value pairs
```
No need of leaving obvious comments.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,220 @@
+
+
+def main():
+    args = parse_args()
```
Main function is in complete violation of DRY & SRP Principle.

ahmed-buksh replied:
Another unaddressed comment.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,220 @@
+    mean_humidity = []
+
+    for day_data in weather_data:
+        if day_data['Max TemperatureC']:
```
It is recommended to use double quotes.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,220 @@
+
+
+def calculate_stats_month(weather_data):
+    high_temps = []
```
What does `temps` define? Try to be descriptive with naming.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,220 @@
+        next(file)
+
+        for line in file:
+            line = line.strip()
+
+            values = line.split(',')
```
Can we find a better way of reading? Going line by line and splitting on commas will be very costly computationally.

ahmed-buksh replied:
Unaddressed comment!

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,220 @@
+            next(file)
+
+            for line in file:
+                line = line.strip()
+
+                values = line.split(',')
```
Violation of DRY Principle. We are doing same work in another function for reading.

ahmed-buksh reviewed (changes requested):
- Look into SRP & DRY Principle
- Make sure to use descriptive name for variables and functions
- No need of leaving obvious comments in your code

usman-altaf-co committed:
Update weather.py

ahmed-buksh reviewed (changes requested):
- Look into SRP & DRY Principle
- Make sure to use descriptive name for variables and functions
- No need of leaving obvious comments in your code

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,212 @@
+    'Precipitationmm', 'CloudCover', 'Events', 'WindDirDegrees'
+]
+
+month_abbr = dict(zip(range(1, 13), calendar.month_abbr[1:]))
```
Can't we directly get months from some utility? Why we need to pass range?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,212 @@
+
+
+def load_weather_data(file_path, year):
+    weather_data = []
```
Data is an ambiguous word. Try to be precise with your variable and function naming.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,212 @@
+    weather_data = []
+
+    for month_abbr in months:
+        file_name = f"Murree_weather_{year}_{month_abbr}.txt"
```
Which files are we talking about?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,212 @@
+month_abbr = dict(zip(range(1, 13), calendar.month_abbr[1:]))
+
+
+def load_weather_data(file_path, year):
```
This function is in violation of SRP Principle. We need to ensure that our all functions are in complete compliance with SRP.

ahmed-buksh replied:
Name of parameters also requires a reconsideration.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,212 @@
+        with open(full_path, mode='r') as file:
+            next(file)
+
+            for line in file:
```
Vague names.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,212 @@
+
+    months = month_abbr
+
+    file_name = f"Murree_weather_{year}_{months[month]}.txt"
+    full_path = os.path.join(file_path, file_name)
+
+    if not os.path.isfile(full_path):
+        print('Year or Months Data does not exist')
+
+    with open(full_path, mode='r') as file:
+        next(file)
+
+        for line in file:
+            line = line.strip()
```
Violation of DRY Principle. We have two functions doing same work.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,212 @@
+            temp = int(day_data["Max TemperatureC"])
+            if temp > highest_temp:
+                highest_temp = temp
+                highest_temp_day = day_data["PKT"]
```
We do have `PKST` key in some cases. Have you tested your code thoroughly?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,212 @@
+    avg_mean_humidity = sum(mean_humidity) / len(mean_humidity)
+
+    return {
+        'avgHighest': avg_high_temp,
+        'avgLowest': avg_low_temp,
+        'avgMeanHumidity': avg_mean_humidity
```
As mentioned before it's recommended to use double quotes.

ahmed-buksh reviewed (changes requested):
- Many previous comments are still unaddressed
- Naming of variables and function is below par
- Suggestions are made on one point but it is expected that you will be fixing it on all other instances of your code

usman-altaf-co committed:
Iteration-2

- Data loading improved
- Followed DRY and SRP
- Variables, functions and parameter names improved
- PKST key also being handled now.
- Double Quotation used, instead of single.
- Main Function improved

usman-altaf-co committed:
Update weather.py

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,198 @@
+    "PKT", "PKST", "Max TemperatureC", "Min TemperatureC", "Mean Humidity"
+]
+
+month_abbr_map = {index: month for index, month in enumerate(calendar.month_abbr) if month}
```
Instead of creating map can't we directly consume calendar utilities for getting month name?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,198 @@
+
+
+def read_weather_file(weather_file_path):
+    murree_weather_month = []
```
Does this name looks descriptive to you?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,198 @@
+    with open(weather_file_path, mode="r") as file:
+        csv_reader = csv.DictReader(file)
+        headers = [header.strip() for header in csv_reader.fieldnames]
+        for row in csv_reader:
```
What does row defines? Be descriptive with your names.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,198 @@
+
+        if daily_data["Max TemperatureC"]:
+            max_temp = int(daily_data["Max TemperatureC"])
+            if max_temp > highest_temperature:
+                highest_temperature = max_temp
+                highest_temprature_day = daily_data[date_key]
```
Max temp calculation and Max humidity calculation are in violation of DRY Principle. We can create a function and on basis of key make it work.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,198 @@
+        minString = "".join(min_str)
+
+        if max_temprature != 0:
+            # print(f"{day} {RED}{maxString}{RESET} {max_temp}C")
+            # print(f"{day} {BLUE}{minString}{RESET} {min_temp}C")
+
+            # Bonus Task
```
Commented code should be removed before review submission.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,198 @@
+    print(f"Highest: {high_temprature}C on {month_abbr_map[int(date_high[1])]} {date_high[2]}")
+    print(f"Lowest: {low_temprature}C on {month_abbr_map[int(date_low[1])]} {date_low[2]}")
+    print(f"Humidity: {humid_val}% on {month_abbr_map[int(date_humid[1])]} {date_humid[2]}")
+    print("\n")
```
No need to handle line break.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,198 @@
+    }
+
+
+def calculate_monthly_stats(weather_data):
```
Is this function calculating stats or averages?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,198 @@
+import argparse
+import os
+import calendar
+import csv
```
Kindly look into PEP8 recommendation of writing imports.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,198 @@
+    highest_humidity_day = None
+
+    for daily_data in weather_data:
+        date_key = "PKT" if "PKT" in daily_data else "PKST"
```
Can we handle date_key at the time of retrieval by using get with dict?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,198 @@
+    low_temperatures = []
+    mean_humidities = []
+
+    for daily_data in weather_data:
+        if daily_data["Max TemperatureC"]:
+            high_temperatures.append(int(daily_data["Max TemperatureC"]))
+
+        if daily_data["Min TemperatureC"]:
+            low_temperatures.append(int(daily_data["Min TemperatureC"]))
+
+        if daily_data["Mean Humidity"]:
+            mean_humidities.append(int(daily_data["Mean Humidity"]))
```
Average calculation should be on generic function which is receiving list of inputs and calculate averages on it with respect to key.
Currently this is violation of DRY Principle.

ahmed-buksh reviewed (changes requested):
- Variable naming requires improvement
- Functions should be in compliance with SRP & DRY Principle

usman-altaf-co committed:
Update weather.py

- Improved SRP and DRY implementation
- Improved Variable and function naming

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,182 @@
+                key.strip(): value
+                for key, value in daily_weather_record.items()
+            }
+            filtered_daily_record = {
```
`filtered_daily_record ` --> `filtered_daily_weather_record`

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,182 @@
+def load_yearly_weather_data(directory_path, year):
+    year_weather_data = []
+
+    for month in range(1, 13):
```
Why we are hardcoding indexes to generate month name by ranges? Can we explore `calendar` library?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,182 @@
+    return read_weather_file(full_file_path)
+
+
+def calculate_extremes(key, weather_data, comparison, extreme_value, extreme_day):
```
Ideally a function should have one parameter. We can take it to two but max three and in case of three we should be answerable that what was the reason of having these much parameters. Look into clean code recommendation of function parameters.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,182 @@
+    lowest_temprature_day = None
+    highest_humidity_day = None
+
+    highest_temperature, highest_temprature_day = calculate_extremes(
+        "Max TemperatureC", weather_data, lambda x,y : x > y, highest_temperature, highest_temprature_day)
```
```
highest_temperature, highest_temprature_day = calculate_extremes(
    "Max TemperatureC",
    weather_data,
    lambda x,y : x > y, highest_temperature,
    highest_temprature_day
)
```

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,182 @@
+        max_temp_bar , min_temp_bar, max_temprature, min_temprature = create_temperature_bars(day_data)
+        day = day_data.get("PKT", day_data.get("PKST")).split("-")[2]
+
+        if max_temprature != 0:
```
For some cities in world max temperature can be 0, what is the rationale of such check?

ahmed-buksh replied:
Kindly look into spellings carefully.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,182 @@
+def create_temperature_bars(day_data):
+    max_temprature = 0
+    min_temprature = 0
+    if day_data["Max TemperatureC"]:
```
What does `day_data` defines? Variable names should be descriptive and self-explanatory.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,182 @@
+    low_temprature, low_day = stats["Lowest"]
+    humid_val, humid_day = stats["Humidity"]
+
+    date_high = high_day.split("-")
+    date_low = low_day.split("-")
+    date_humid = humid_day.split("-")
```
Can we use better and meaningful names?

ahmed-buksh reviewed (changes requested):
- Do ensure that there are no spelling mistakes
- Variable and function names require a reconsideration
- Do add docstrings to all functions

usman-altaf-co committed:
Update weather.py

- Added docstrings to each function
- Variable and function naming improved.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,344 @@
+    if isinstance(month, int):
+        month_abbr = calendar.month_abbr[month]
+    else:
+        month_abbr = month
```
It can be a simple one liner.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,344 @@
+      Keys: "Highest", "Lowest", "Humidity".
+      Values: Tuples (extreme_value, extreme_day) for each category.
+    """
+    def calculate_extremes(key, comparison):
```
Any specific reasons for such closures?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,344 @@
+    print(
+        f"Highest: {highest_temperature}C on "
+        f"{calendar.month_abbr[int(highest_temperature_day[1])]} "
+        f"{highest_temperature_day[2]}"
```
We need to avoid hardcoded index approaches.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,344 @@
+    print(calendar.month_abbr[month], year)
+    for daily_weather_record in month_weather_data:
+        (max_temperature_bar, min_temperature_bar,
+         max_temperature, min_temperature) = create_temperature_bars(daily_weather_record)
```
```
(
    max_temperature_bar,
    min_temperature_bar, 
    max_temperature, 
    min_temperature
) = create_temperature_bars(daily_weather_record)
```

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,344 @@
+    """
+    max_temperature = 0
+    min_temperature = 0
+    if day_weather_record["Max TemperatureC"]:
```
Why we are only checking Max Temperature only?

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,344 @@
+    - argparse.Namespace: Parsed arguments object containing parsed arguments
+      as attributes.
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument("directory", type=str)
+    parser.add_argument("-e", "--year", type=int)
+    parser.add_argument("-a", "--year_month", type=str)
+    parser.add_argument("-c", "--chart", type=str)
+    return parser.parse_args()
```
Blank lines should be used to increase readability of code
```
parser = argparse.ArgumentParser()

parser.add_argument("directory", type=str)
parser.add_argument("-e", "--year", type=int)
parser.add_argument("-a", "--year_month", type=str)
parser.add_argument("-c", "--chart", type=str)

return parser.parse_args()
```

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,344 @@
+    return month_weather_data
+
+
+def load_year_weather_data(weather_files_directory, year):
+    """
+    Loads weather data for a specified year from multiple monthly files.
+
+    Args:
+    - weather_files_directory (str): Directory path where weather files are stored.
+    - year (int): Year for which weather data is to be loaded.
+
+    Returns:
+    - list: Combined list of weather records for the entire year.
+    """
+    year_weather_data = []
+
+    for month_abbr in calendar.month_abbr[1:]:
+        month_weather_data = load_month_weather_data(weather_files_directory, year, month_abbr)
+        year_weather_data.extend(month_weather_data)
+
+    return year_weather_data
+
+
+def load_month_weather_data(weather_files_directory, year, month):
```
Loading data should be irrespective of month or year. It should be dependent on size of input based on year or month. We need to make it abstract.

ahmed-buksh commented on weather.py:
```diff
@@ -0,0 +1,344 @@
+    if not os.path.isfile(weather_data_file):
+        return month_weather_data
+
+    with open(weather_data_file, mode="r") as file:
```
Which `file` are we talking about?

ahmed-buksh reviewed (changes requested):
- Let's move towards class based approach. Make sure to have individual classes for all concerns in different files

usman-altaf-co committed:
Class Based Approach

ahmed-buksh commented on main.py:
```diff
@@ -0,0 +1,10 @@
+
+
+def main():
+    weather_data_processor = WeatherDataHandler()
```
What differentiates processor from a handler?

ahmed-buksh commented on weather_analyzer.py:
```diff
@@ -0,0 +1,120 @@
+class WeatherDataAnalyzer:
+    @staticmethod
+    def calculate_extremes(year_weather_data, key, comparison):
```
Which extremes are we talking about?

ahmed-buksh replied:
All of your parameter names are ambiguous, rename them. As per clean code, number of parameters should be minimum to a function ideally 1. In case of more than 2 parameters, one should be seriously answerable of what they are doing.

ahmed-buksh commented on weather_analyzer.py:
```diff
@@ -0,0 +1,120 @@
+        Returns:
+        - tuple: Tuple containing the extreme value and the corresponding day.
+        """
+        if key == "Max TemperatureC":
+            extreme_value = float("-inf")
+        elif key == "Min TemperatureC":
+            extreme_value = float("inf")
```
Why do we require to set these base values for calculating extremes? If we have a list of values, can't we simply ask for max or min from it?

ahmed-buksh commented on weather_analyzer.py:
```diff
@@ -0,0 +1,120 @@
+          Keys: "Highest", "Lowest", "Humidity".
+          Values: Tuples (extreme_value, extreme_day) for each category.
+        """
+        (highest_temperature,
+         highest_temperature_day) = WeatherDataAnalyzer.calculate_extremes(
+            year_weather_data, "Max TemperatureC", lambda x, y: x > y
+        )
```
```
(highest_temperature,highest_temperature_day) = WeatherDataAnalyzer.calculate_extremes(
            year_weather_data,
            "Max TemperatureC",
            lambda x, y: x > y
)
```

ahmed-buksh replied:
Ensure such reformatting approaches all across your code.

ahmed-buksh commented on weather_analyzer.py:
```diff
@@ -0,0 +1,120 @@
+        - dict: Dictionary containing the calculated averages.
+          Keys: "Average Highest Temperature", "Average Lowest Temperature", "Average Mean Humidity".
+        """
+        avg_high_temperature = (WeatherDataAnalyzer.calculate_average
```
Kindly avoid abbreviations as it leads to confusion.

ahmed-buksh commented on weather_data.py:
```diff
@@ -0,0 +1,75 @@
+class WeatherDataLoader:
+    def __init__(self, directory):
+        self.directory = directory
+        self.required_labels = [
+            "PKT", "PKST", "Max TemperatureC",
+            "Min TemperatureC", "Mean Humidity"
+        ]
```
```
self.required_labels = [
    "PKT",
    "PKST",
    "Max TemperatureC",
    "Min TemperatureC",
    "Mean Humidity",
]
```

ahmed-buksh commented on weather_data.py:
```diff
@@ -0,0 +1,75 @@
+        Returns:
+        - list: Combined list of weather records for the specified time period.
+        """
+        if month:
+            month_abbr = calendar.month_abbr[month]
+            return self.read_weather_file(year, month_abbr)
+        else:
+            year_weather_data = []
+
+            for month_abbr in calendar.month_abbr[1:]:
+                month_weather_data = self.read_weather_file(year, month_abbr)
+                year_weather_data.extend(month_weather_data)
+
+            return year_weather_data
```
Why does loading needs to work differently for a month or year? If it's month, one file should be loaded and in case of year it has to be twelve. Why we are catering it in different way?

ahmed-buksh commented on weather_display.py:
```diff
@@ -0,0 +1,130 @@
+        )
+
+    @staticmethod
+    def show_month_averages(month_weather_averages):
```
This method is displaying or printing. Show is a vague word which is not revealing the right intent.

ahmed-buksh commented on weather_display.py:
```diff
@@ -0,0 +1,130 @@
+             max_temperature,
+             min_temperature) = (WeatherDataPresenter.
+                                 create_temperature_bars(daily_weather_record))
+            day = daily_weather_record.get("PKT",
+                                           daily_weather_record.get("PKST")).split("-")[2]
```
Reformat it.

ahmed-buksh reviewed (changes requested):
- Naming requires serious reconsideration of all names
- Indentation is not properly done, kindly follow provided suggestions
- Rethink the handling and reading of data in more logical way

usman-altaf-co committed:
- Improved with changes

Classes

ahmed-buksh commented on weather_analyzer.py:
```diff
@@ -0,0 +1,127 @@
+class WeatherDataAnalyzer:
+    @staticmethod
+    def set_year_weather_readings(loaded_year_weather_readings):
+        global year_weather_readings
```
Any specific reason for making it global?

ahmed-buksh commented on weather_analyzer.py:
```diff
@@ -0,0 +1,127 @@
+year_weather_readings = None
```
Can't we make it a class variable?

ahmed-buksh commented on weather_analyzer.py:
```diff
@@ -0,0 +1,127 @@
+
+        for daily_weather_record in year_weather_readings:
+            if daily_weather_record[weather_attribute]:
+                if int(daily_weather_record[weather_attribute]) == extreme_weather_value:
```
Can't we use `and` for chaining all conditions?

ahmed-buksh commented on weather_analyzer.py:
```diff
@@ -0,0 +1,127 @@
+            "Highest": (highest_temperature, highest_temperature_day),
+            "Lowest": (lowest_temperature, lowest_temperature_day),
+            "Humidity": (highest_humidity, highest_humidity_day)
+        }
```
Why we don't use analyzer function directly in return statement?

ahmed-buksh commented on weather_analyzer.py:
```diff
@@ -0,0 +1,127 @@
+          weather_attributes: "Average Highest Temperature", "Average Lowest Temperature", "Average Mean Humidity".
+        """
+        average_high_temperature = (WeatherDataAnalyzer.calculate_average
+                                    (month_weather_readings, "Max TemperatureC"))
```
```
average_high_temperature = (
    WeatherDataAnalyzer.calculate_average
    (month_weather_readings, "Max TemperatureC")
)
```

ahmed-buksh commented on weather_data_processor.py:
```diff
@@ -0,0 +1,44 @@
+        self.weather_data_loader = WeatherDataLoader(args.directory)
+
+        if args.year:
+            year_weather_data = (self.weather_data_loader.load_weather_readings(args.year))
+
+            WeatherDataAnalyzer.set_year_weather_readings(year_weather_data)
+
+            weather_extremes = (WeatherDataAnalyzer.calculate_year_weather_extremes())
+
+            WeatherDataPresenter.display_year_weather_extremes(weather_extremes)
```
Any particular reason for these blank lines?

ahmed-buksh commented on weather_display.py:
```diff
@@ -0,0 +1,132 @@
+          Keys: "Highest", "Lowest", "Humidity".
+          Values: Tuples (extreme_value, extreme_day) for each category.
+        """
+        (highest_temperature,
+         highest_temperature_day) = weather_extremes["Highest"]
```
```
(highest_temperature, highest_temperature_day) = weather_extremes["Highest"]
```
It is not exceeding line length limit then what is the point of splitting it?

ahmed-buksh commented on weather_display.py:
```diff
@@ -0,0 +1,132 @@
+        lowest_temperature_day = lowest_temperature_day.split("-")
+        most_humid_day = most_humid_day.split("-")
+
+        highest_temperature_month = highest_temperature_day[1]
+        lowest_temperature_month = lowest_temperature_day[1]
+        most_humid_month = most_humid_day[1]
+
+        highest_temperature_day = highest_temperature_day[2]
+        lowest_temperature_day = lowest_temperature_day[2]
+        most_humid_day = most_humid_day[2]
```
As mentioned before we are supposed to avoid hardcoded indexes!

usman-altaf-co committed:
Improved with changes
